haystack_experimental/core/pipeline/async_pipeline.py

```diff
@@ -2,17 +2,592 @@
+        cached_receivers = {
+            n: self._find_receivers_from(n) for n in ordered_names
+        }
+
+        # We fill the queue once and raise if all components are BLOCKED
+        self.validate_pipeline(self._fill_queue(ordered_names, inputs_state))
+
+        # Single parent span for entire pipeline execution
+        with tracing.tracer.trace(
+            "haystack.async_pipeline.run",
+            tags={
+                "haystack.pipeline.input_data": data,
+                "haystack.pipeline.metadata": self.metadata,
+                "haystack.pipeline.max_runs_per_component": self._max_runs_per_component,
+            },
+        ) as parent_span:
+
+            # -------------------------------------------------
+            # We define some functions here so that they have access to local runtime state
+            # (inputs, tasks, scheduled components) via closures.
+            # -------------------------------------------------
+            async def _run_component_async(
+                component_name: str,
+                component_inputs: Dict[str, Any],
+            ) -> Dict[str, Any]:
+                """
+                Runs one component.
+
+                If the component supports async, await directly it will run async; otherwise offload to executor.
+                Updates visits count, writes outputs to `inputs_state`,
+                and returns pruned outputs that get stored in `pipeline_outputs`.
+
+                :param component_name: The name of the component.
+                :param component_inputs: Inputs for the component.
+                :returns: Outputs from the component that can be yielded from run_async_generator.
+                """
+                comp_dict = self._get_component_with_graph_metadata(component_name)
+                if comp_dict["visits"] >= self._max_runs_per_component:
+                    raise PipelineMaxComponentRuns(
+                        f"Max runs for '{component_name}' reached."
+                    )
+
+                instance: Component = comp_dict["instance"]
+                with tracing.tracer.trace(
+                    "haystack.component.run_async", parent_span=parent_span
+                ):
+                    logger.info("Running component {name}", name=component_name)
+
+                    if getattr(instance, "__haystack_supports_async__", False):
+                        outputs = await instance.run_async(**component_inputs)  # type: ignore
+                    else:
+                        loop = asyncio.get_event_loop()
+                        outputs = await loop.run_in_executor(
+                            None, lambda: instance.run(**component_inputs)
+                        )
+
+                comp_dict["visits"] += 1
+
+                if not isinstance(outputs, dict):
+                    raise PipelineRuntimeError(
+                        f"Component '{component_name}' returned non-dict outputs."
+                    )
```

Comment:
Suggestions to improve the error message to provide more meaningful feedback for debugging.
```suggestion
                if not isinstance(outputs, dict):
    raise PipelineRuntimeError(
        f"Component '{component_name}' returned an invalid output type. "
        f"Expected a dict, but got {type(outputs).__name__} instead. "
    )
```